WEEK 1: Data Collection, Environment Setup & Data Cleaning

In [1]:
import pandas as pd
import os

In [2]:
print(os.getcwd())

C:\Users\INDIA\OneDrive\Desktop\Infotact_Omnichannel_Retail_Analytics\Dataset\archive (1)


In [3]:
print(os.listdir())

['.ipynb_checkpoints', 'dim_brand.parquet', 'dim_campaign.parquet', 'dim_category.parquet', 'dim_customer.parquet', 'dim_date.parquet', 'dim_location.parquet', 'dim_product.parquet', 'dim_promotion.parquet', 'dim_store.parquet', 'dim_subcategory.parquet', 'fact_clickstream.parquet', 'fact_sale.parquet', 'fact_transaction.parquet', 'inventory.parquet', 'Untitled.ipynb']


In [4]:
sales = pd.read_parquet("fact_sale.parquet")
sales.head()

,sale_id,transaction_id,session_id,transaction_timestamp,transaction_date_id,product_id,quantity,unit_cost,unit_price,line_cost,line_total,aov_category
0,12000363,8000125,9000924.0,2023-07-09 14:24:59,20230709,335,1,20.00,24.25,20.00,24.25,Low
1,12000364,8000125,9000924.0,2023-07-09 14:24:59,20230709,440,1,98.00,131.09,98.00,131.09,Low
2,12000365,8000126,9000930.0,2023-08-05 00:34:50,20230805,165,1,20.00,24.17,20.00,24.17,Low
3,12000366,8000126,9000930.0,2023-08-05 00:34:50,20230805,135,1,230.00,277.14,230.00,277.14,Low
4,12000367,8000126,9000930.0,2023-08-05 00:34:50,20230805,399,1,140.00,167.43,140.00,167.43,Low


In [5]:
sales.shape

(1799836, 12)

In [6]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1799836 entries, 0 to 1799835
Data columns (total 12 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   sale_id                int32         
 1   transaction_id         int32         
 2   session_id             float64       
 3   transaction_timestamp  datetime64[us]
 4   transaction_date_id    int32         
 5   product_id             int32         
 6   quantity               int32         
 7   unit_cost              object        
 8   unit_price             object        
 9   line_cost              object        
 10  line_total             object        
 11  aov_category           object        
dtypes: datetime64[us](1), float64(1), int32(5), object(5)
memory usage: 130.5+ MB


In [7]:
sales.columns.tolist()

['sale_id',
 'transaction_id',
 'session_id',
 'transaction_timestamp',
 'transaction_date_id',
 'product_id',
 'quantity',
 'unit_cost',
 'unit_price',
 'line_cost',
 'line_total',
 'aov_category']

In [8]:
sales.isnull().sum()

sale_id                       0
transaction_id                0
session_id               679257
transaction_timestamp         0
transaction_date_id           0
product_id                    0
quantity                      0
unit_cost                     0
unit_price                    0
line_cost                     0
line_total                    0
aov_category                  0
dtype: int64

In [9]:
sales.duplicated().sum()

np.int64(0)

In [10]:
sales.describe()

,sale_id,transaction_id,session_id,transaction_timestamp,transaction_date_id,product_id,quantity
count,1.799836e+06,1.799836e+06,1.120579e+06,1799836,1.799836e+06,1.799836e+06,1799836.0
mean,1.290028e+07,8.450064e+06,1.728324e+07,2024-11-19 03:38:07.403819,2.024407e+07,2.327813e+02,1.0
min,1.200036e+07,8.000125e+06,9.000864e+06,2023-01-01 00:26:35,2.023010e+07,1.000000e+00,1.0
25%,1.245032e+07,8.225082e+06,1.404357e+07,2024-04-16 09:10:34.750000,2.024042e+07,1.160000e+02,1.0
50%,1.290028e+07,8.450138e+06,1.782423e+07,2025-01-07 16:51:37,2.025011e+07,2.290000e+02,1.0
75%,1.335024e+07,8.674933e+06,2.065788e+07,2025-07-30 08:24:28,2.025073e+07,3.540000e+02,1.0
max,1.380020e+07,8.900124e+06,2.350086e+07,2025-12-31 23:44:00,2.025123e+07,4.700000e+02,1.0
std,5.195680e+05,2.597789e+05,3.993262e+06,NaN,7.558526e+03,1.374438e+02,0.0


# Initial Assessment

Rows: 1,799,836

Columns: 12

Observations:
- To be filled after NULL analysis
- To be filled after duplicate analysis
- To be filled after datatype analysis

In [11]:
customer = pd.read_parquet("dim_customer.parquet")

customer.shape

(150000, 16)

In [12]:
customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   customer_id              150000 non-null  int32         
 1   email_address            150000 non-null  object        
 2   first_name               150000 non-null  object        
 3   last_name                150000 non-null  object        
 4   gender                   150000 non-null  object        
 5   customer_persona         150000 non-null  object        
 6   birth_date               150000 non-null  object        
 7   birth_year               150000 non-null  int32         
 8   location_id              150000 non-null  int32         
 9   signup_date              150000 non-null  datetime64[us]
 10  signup_date_id           150000 non-null  int32         
 11  signup_channel           150000 non-null  object        
 12  loyalty_status  

In [13]:
customer.head()

,customer_id,email_address,first_name,last_name,gender,customer_persona,birth_date,birth_year,location_id,signup_date,signup_date_id,signup_channel,loyalty_status,estimated_annual_income,email_opt_in,sms_opt_in
0,1,diana.haynes.1@example.com,diana,haynes,Female,Tech Enthusiast,2000-12-31,2000,4,2019-12-23,20191223,Mobile,Elite,61151.00,True,True
1,2,joyce.simon.2@bac.com,joyce,simon,Female,Tech Enthusiast,2000-01-01,2000,1,2021-09-14,20210914,Mobile,Gold,71611.00,False,False
2,3,monique.colon.3@example.com,monique,colon,Female,Everyday Shopper,1976-12-31,1976,3,2008-07-27,20080727,Online,Basic,87003.00,False,False
3,4,elizabeth.heath.4@example.com,elizabeth,heath,Female,Everyday Shopper,1987-01-01,1987,3,2009-10-14,20091014,In-Store,Basic,84305.00,True,True
4,5,pamela.coleman.5@xyz.com,pamela,coleman,Female,Tech Enthusiast,1995-01-01,1995,2,2024-02-26,20240226,Online,Gold,69674.00,True,True


In [14]:
product = pd.read_parquet("dim_product.parquet")

product.shape

(470, 9)

In [15]:
product.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 470 entries, 0 to 469
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   product_id       470 non-null    int32 
 1   product_name     470 non-null    object
 2   category_id      470 non-null    int32 
 3   subcategory_id   470 non-null    int32 
 4   brand_id         470 non-null    int32 
 5   unit_cost        470 non-null    object
 6   unit_price       470 non-null    object
 7   warranty_years   470 non-null    int32 
 8   product_segment  470 non-null    object
dtypes: int32(5), object(4)
memory usage: 24.0+ KB


In [16]:
product.head()

,product_id,product_name,category_id,subcategory_id,brand_id,unit_cost,unit_price,warranty_years,product_segment
0,1,Philips 55PUT7630/98 55'' UHD Smart Led TV,1,1,1,280.00,327.68,2,Mid Tier
1,2,Samsung UE55U8000FUXMU 55'' Led TV,1,1,2,310.00,372.46,2,Mid Tier
2,3,Samsung TU32H5005FKXXC 32'' Led TV,1,1,2,160.00,179.74,2,Entry Level
3,4,Samsung UE43U8000FUXMU 43'' Led TV,1,1,2,230.00,271.20,2,Entry Level
4,5,Philips 40PFT6509/98 40'' Led TV,1,1,1,140.00,169.46,2,Entry Level


In [17]:
store = pd.read_parquet("dim_store.parquet")

store.shape

(50, 8)

In [18]:
store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   store_id            50 non-null     int32         
 1   store_name          50 non-null     object        
 2   location_id         50 non-null     int32         
 3   store_type          50 non-null     object        
 4   store_size          50 non-null     int32         
 5   opening_date        50 non-null     datetime64[us]
 6   opening_date_id     50 non-null     int32         
 7   foot_traffic_index  50 non-null     int32         
dtypes: datetime64[us](1), int32(5), object(2)
memory usage: 2.3+ KB


In [19]:
store.head()

,store_id,store_name,location_id,store_type,store_size,opening_date,opening_date_id,foot_traffic_index
0,1,ELEC-MART Store #1,9,Outlet,774,2001-05-30 10:00:00,20010530,53
1,2,ELEC-MART Store #2,6,Warehouse,3863,2001-12-18 10:00:00,20011218,68
2,3,ELEC-MART Store #3,8,Standalone,1142,2003-01-28 10:00:00,20030128,84
3,4,ELEC-MART Store #4,7,Standalone,963,2003-03-20 10:00:00,20030320,77
4,5,ELEC-MART Store #5,4,Mall,395,2003-04-28 10:00:00,20030428,89


In [20]:
location = pd.read_parquet("dim_location.parquet")

location.shape

(10, 8)

In [21]:
location.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   location_id       10 non-null     int32 
 1   country           10 non-null     object
 2   state_province    10 non-null     object
 3   city              10 non-null     object
 4   location_type     10 non-null     object
 5   location_weight   10 non-null     object
 6   foot_traffic_min  10 non-null     int32 
 7   foot_traffic_max  10 non-null     int32 
dtypes: int32(3), object(5)
memory usage: 648.0+ bytes


In [22]:
location.head()

,location_id,country,state_province,city,location_type,location_weight,foot_traffic_min,foot_traffic_max
0,1,Canada,Ontario,Toronto,Urban,0.27,75,90
1,2,Canada,Ontario,Ottawa,Urban,0.12,75,90
2,3,Canada,Ontario,Mississauga,Suburban,0.09,50,70
3,4,Canada,Quebec,Montreal,Urban,0.20,75,90
4,5,Canada,British Columbia,Vancouver,Urban,0.09,75,90


In [23]:
inventory = pd.read_parquet("inventory.parquet")

inventory.shape

(846000, 10)

In [24]:
inventory.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 846000 entries, 0 to 845999
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   inventory_id    846000 non-null  int32 
 1   product_id      846000 non-null  int32 
 2   store_id        846000 non-null  int32 
 3   snapshot_month  846000 non-null  object
 4   starting_stock  846000 non-null  int32 
 5   received_stock  846000 non-null  int32 
 6   sold_units      846000 non-null  int32 
 7   closing_stock   846000 non-null  int32 
 8   backorder_flag  846000 non-null  bool  
 9   shrinkage_loss  846000 non-null  int32 
dtypes: bool(1), int32(8), object(1)
memory usage: 33.1+ MB


In [25]:
inventory.head()

,inventory_id,product_id,store_id,snapshot_month,starting_stock,received_stock,sold_units,closing_stock,backorder_flag,shrinkage_loss
0,1648889,361,25,2025-01-01,38,0,0,38,False,0
1,1648890,361,25,2025-02-01,38,0,1,37,False,0
2,1648891,361,25,2025-03-01,37,0,1,36,False,0
3,1648892,361,25,2025-04-01,36,1,1,36,False,0
4,1648893,361,25,2025-05-01,36,1,3,34,False,0


In [26]:
brand = pd.read_parquet("dim_brand.parquet")

brand.shape

(46, 2)

In [27]:
brand.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46 entries, 0 to 45
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Brand_ID    46 non-null     int32 
 1   Brand_Name  46 non-null     object
dtypes: int32(1), object(1)
memory usage: 680.0+ bytes


In [28]:
brand.head()

,Brand_ID,Brand_Name
0,1,Philips
1,2,Samsung
2,3,Hisense
3,4,Toshiba
4,5,Sony


In [29]:
category = pd.read_parquet("dim_category.parquet")

category.shape

(10, 2)

In [32]:
category.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   category_id    10 non-null     int32 
 1   category_name  10 non-null     object
dtypes: int32(1), object(1)
memory usage: 248.0+ bytes


In [33]:
category.head()

,category_id,category_name
0,1,Televisions
1,2,Computing
2,3,Home Cinema
3,4,Cameras
4,5,Home Security


In [34]:
subcategory = pd.read_parquet("dim_subcategory.parquet")

subcategory.shape

(29, 3)

In [35]:
subcategory.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29 entries, 0 to 28
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   subcategory_id    29 non-null     int32 
 1   subcategory_name  29 non-null     object
 2   category_id       29 non-null     int32 
dtypes: int32(2), object(1)
memory usage: 592.0+ bytes


In [36]:
subcategory.head()

,subcategory_id,subcategory_name,category_id
0,1,TVs,1
1,2,TV Accessories,1
2,3,Laptops,2
3,4,Desktops,2
4,5,Monitors,2


In [37]:
date = pd.read_parquet("dim_date.parquet")

date.shape

(9479, 13)

In [38]:
date.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9479 entries, 0 to 9478
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   date_id         9479 non-null   int32 
 1   date            9479 non-null   object
 2   year            9479 non-null   int32 
 3   quarter         9479 non-null   int32 
 4   month           9479 non-null   int32 
 5   month_name      9479 non-null   object
 6   day             9479 non-null   int32 
 7   week_of_year    9479 non-null   int32 
 8   day_of_week     9479 non-null   int32 
 9   day_name        9479 non-null   object
 10  is_weekend      9479 non-null   bool  
 11  is_month_start  9479 non-null   bool  
 12  is_month_end    9479 non-null   bool  
dtypes: bool(3), int32(7), object(3)
memory usage: 509.3+ KB


In [39]:
date.head()

,date_id,date,year,quarter,month,month_name,day,week_of_year,day_of_week,day_name,is_weekend,is_month_start,is_month_end
0,20010530,2001-05-30,2001,2,5,May,30,22,3,Wednesday,False,False,False
1,20010531,2001-05-31,2001,2,5,May,31,22,4,Thursday,False,False,True
2,20010601,2001-06-01,2001,2,6,June,1,22,5,Friday,False,True,False
3,20010602,2001-06-02,2001,2,6,June,2,22,6,Saturday,True,False,False
4,20010603,2001-06-03,2001,2,6,June,3,22,0,Sunday,True,False,False


In [40]:
promotion = pd.read_parquet("dim_promotion.parquet")

promotion.shape

(150, 13)

In [41]:
promotion.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   promo_id             150 non-null    int32         
 1   promo_name           150 non-null    object        
 2   promo_type           150 non-null    object        
 3   discount_type        150 non-null    object        
 4   discount_value       150 non-null    object        
 5   promo_start_date     150 non-null    datetime64[us]
 6   promo_start_date_id  150 non-null    int32         
 7   promo_end_date       150 non-null    datetime64[us]
 8   promo_end_date_id    150 non-null    int32         
 9   promo_duration       150 non-null    int32         
 10  promo_code           150 non-null    object        
 11  is_active            150 non-null    bool          
 12  promo_description    150 non-null    object        
dtypes: bool(1), datetime64[us](2), int3

In [42]:
promotion.head()

,promo_id,promo_name,promo_type,discount_type,discount_value,promo_start_date,promo_start_date_id,promo_end_date,promo_end_date_id,promo_duration,promo_code,is_active,promo_description
0,1,Weekend Flash Sale,Flash,Fixed_Amount_Discount,25.00,2023-12-10,20231210,2023-12-11,20231211,1,FLA-202312-1,False,Weekend Flash Sale Get $25 Off
1,2,Mobile Life Sale,Major,Percentage_Discount,0.10,2023-07-24,20230724,2023-08-07,20230807,14,MAJ-202307-2,False,Mobile Life Sale Get 10% Off
2,3,Customer Appreciation Sale,Major,Fixed_Amount_Discount,25.00,2023-07-15,20230715,2023-07-29,20230729,14,MAJ-202307-3,False,Customer Appreciation Sale Get $25 Off
3,4,Limited Time Offers,Medium,Fixed_Amount_Discount,20.00,2023-12-31,20231231,2024-01-07,20240107,7,MED-202312-4,False,Limited Time Offers Get $20 Off
4,5,New Year Mega Sale,Major,Percentage_Discount,0.05,2023-02-21,20230221,2023-03-07,20230307,14,MAJ-202302-5,False,New Year Mega Sale Get 5% Off


In [43]:
campaign = pd.read_parquet("dim_campaign.parquet")

campaign.shape

(120, 8)

In [44]:
campaign.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   campaign_id             120 non-null    int32         
 1   campaign_name           120 non-null    object        
 2   campaign_channel        120 non-null    object        
 3   promo_id                80 non-null     float64       
 4   campaign_start_date     120 non-null    datetime64[us]
 5   campaign_start_date_id  120 non-null    int32         
 6   campaign_end_date       120 non-null    datetime64[us]
 7   campaign_end_date_id    120 non-null    int32         
dtypes: datetime64[us](2), float64(1), int32(3), object(2)
memory usage: 6.2+ KB


In [45]:
campaign.head()

,campaign_id,campaign_name,campaign_channel,promo_id,campaign_start_date,campaign_start_date_id,campaign_end_date,campaign_end_date_id
0,1,Campaign #1,Email,16.0,2023-09-26,20230926,2023-09-26 03:00:00,20230926
1,2,Campaign #2,Email,NaN,2023-01-12,20230112,2023-01-13 00:00:00,20230113
2,3,Campaign #3,SMS,NaN,2023-09-11,20230911,2023-09-11 12:00:00,20230911
3,4,Campaign #4,Email,19.0,2023-02-09,20230209,2023-02-09 12:00:00,20230209
4,5,Campaign #5,SMS,NaN,2023-12-15,20231215,2023-12-16 00:00:00,20231216


## Data Cleaning

In [46]:
sales.isnull().sum()

sale_id                       0
transaction_id                0
session_id               679257
transaction_timestamp         0
transaction_date_id           0
product_id                    0
quantity                      0
unit_cost                     0
unit_price                    0
line_cost                     0
line_total                    0
aov_category                  0
dtype: int64

In [47]:
inventory.isnull().sum()

inventory_id      0
product_id        0
store_id          0
snapshot_month    0
starting_stock    0
received_stock    0
sold_units        0
closing_stock     0
backorder_flag    0
shrinkage_loss    0
dtype: int64

In [48]:
customer.isnull().sum()

customer_id                0
email_address              0
first_name                 0
last_name                  0
gender                     0
customer_persona           0
birth_date                 0
birth_year                 0
location_id                0
signup_date                0
signup_date_id             0
signup_channel             0
loyalty_status             0
estimated_annual_income    0
email_opt_in               0
sms_opt_in                 0
dtype: int64

In [49]:
product.isnull().sum()

product_id         0
product_name       0
category_id        0
subcategory_id     0
brand_id           0
unit_cost          0
unit_price         0
warranty_years     0
product_segment    0
dtype: int64

In [50]:
store.isnull().sum()

store_id              0
store_name            0
location_id           0
store_type            0
store_size            0
opening_date          0
opening_date_id       0
foot_traffic_index    0
dtype: int64

In [51]:
location.isnull().sum()

location_id         0
country             0
state_province      0
city                0
location_type       0
location_weight     0
foot_traffic_min    0
foot_traffic_max    0
dtype: int64

In [52]:
brand.isnull().sum()

Brand_ID      0
Brand_Name    0
dtype: int64

In [53]:
category.isnull().sum()

category_id      0
category_name    0
dtype: int64

In [54]:
subcategory.isnull().sum()

subcategory_id      0
subcategory_name    0
category_id         0
dtype: int64

In [55]:
date.isnull().sum()

date_id           0
date              0
year              0
quarter           0
month             0
month_name        0
day               0
week_of_year      0
day_of_week       0
day_name          0
is_weekend        0
is_month_start    0
is_month_end      0
dtype: int64

In [56]:
promotion.isnull().sum()

promo_id               0
promo_name             0
promo_type             0
discount_type          0
discount_value         0
promo_start_date       0
promo_start_date_id    0
promo_end_date         0
promo_end_date_id      0
promo_duration         0
promo_code             0
is_active              0
promo_description      0
dtype: int64

In [57]:
campaign.isnull().sum()

campaign_id                0
campaign_name              0
campaign_channel           0
promo_id                  40
campaign_start_date        0
campaign_start_date_id     0
campaign_end_date          0
campaign_end_date_id       0
dtype: int64

In [58]:
sales.duplicated().sum()
inventory.duplicated().sum()
customer.duplicated().sum()
product.duplicated().sum()
store.duplicated().sum()
location.duplicated().sum()
brand.duplicated().sum()
category.duplicated().sum()
subcategory.duplicated().sum()
date.duplicated().sum()
promotion.duplicated().sum()
campaign.duplicated().sum()

np.int64(0)

In [59]:
sales["unit_cost"] = pd.to_numeric(sales["unit_cost"])
sales["unit_price"] = pd.to_numeric(sales["unit_price"])
sales["line_cost"] = pd.to_numeric(sales["line_cost"])
sales["line_total"] = pd.to_numeric(sales["line_total"])

In [60]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1799836 entries, 0 to 1799835
Data columns (total 12 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   sale_id                int32         
 1   transaction_id         int32         
 2   session_id             float64       
 3   transaction_timestamp  datetime64[us]
 4   transaction_date_id    int32         
 5   product_id             int32         
 6   quantity               int32         
 7   unit_cost              float64       
 8   unit_price             float64       
 9   line_cost              float64       
 10  line_total             float64       
 11  aov_category           object        
dtypes: datetime64[us](1), float64(5), int32(5), object(1)
memory usage: 130.5+ MB


## Saving Cleaned Dataset

In [61]:
sales.to_parquet("cleaned_fact_sale.parquet", index=False)
customer.to_parquet("cleaned_dim_customer.parquet", index=False)
product.to_parquet("cleaned_dim_product.parquet", index=False)
store.to_parquet("cleaned_dim_store.parquet", index=False)
inventory.to_parquet("cleaned_inventory.parquet", index=False)

### Tasks Completed

- Downloaded and organized the Omnichannel Retail Analytics dataset.
- Set up the project structure and Jupyter Notebook environment.
- Loaded all fact and dimension tables into Python.
- Explored dataset structure using shape, info(), columns, and head().
- Performed initial data quality assessment by checking missing values and duplicate records.
- Corrected incorrect data types in the sales dataset.
- Saved cleaned datasets for further analysis.

### Outcome

The dataset has been successfully explored, validated, and prepared for SQL database implementation in Week 2.